receiver/dockerstatsreceiver/receiver.go

```diff
@@ -88,29 +92,210 @@ func (r *receiver) scrape(ctx context.Context) (pmetric.Metrics, error) {
+		r.mb.RecordContainerNetworkIoUsageRxErrorsDataPoint(now, int64(stats.RxErrors), netInterface)
+		r.mb.RecordContainerNetworkIoUsageTxErrorsDataPoint(now, int64(stats.TxErrors), netInterface)
+	}
+}
+
+func (r *receiver) recordCPUMetrics(now pcommon.Timestamp, cpuStats *dtypes.CPUStats, prevStats *dtypes.CPUStats) {
+	r.mb.RecordContainerCPUUsageSystemDataPoint(now, int64(cpuStats.SystemUsage))
+	r.mb.RecordContainerCPUUsageTotalDataPoint(now, int64(cpuStats.CPUUsage.TotalUsage))
+	r.mb.RecordContainerCPUUsageKernelmodeDataPoint(now, int64(cpuStats.CPUUsage.UsageInKernelmode))
+	r.mb.RecordContainerCPUUsageUsermodeDataPoint(now, int64(cpuStats.CPUUsage.UsageInUsermode))
+	r.mb.RecordContainerCPUThrottlingDataThrottledPeriodsDataPoint(now, int64(cpuStats.ThrottlingData.ThrottledPeriods))
+	r.mb.RecordContainerCPUThrottlingDataPeriodsDataPoint(now, int64(cpuStats.ThrottlingData.Periods))
+	r.mb.RecordContainerCPUThrottlingDataThrottledTimeDataPoint(now, int64(cpuStats.ThrottlingData.ThrottledTime))
+	r.mb.RecordContainerCPUPercentDataPoint(now, calculateCPUPercent(prevStats, cpuStats))
+
+	for coreNum, v := range cpuStats.CPUUsage.PercpuUsage {
+		r.mb.RecordContainerCPUUsagePercpuDataPoint(now, int64(v), fmt.Sprintf("cpu%s", strconv.Itoa(coreNum)))
+	}
+}
+
+// From container.calculateCPUPercentUnix()
+// https://github.com/docker/cli/blob/dbd96badb6959c2b7070664aecbcf0f7c299c538/cli/command/container/stats_helpers.go
+// Copyright 2012-2017 Docker, Inc.
+// This product includes software developed at Docker, Inc. (https://www.docker.com).
+// The following is courtesy of our legal counsel:
+// Use and transfer of Docker may be subject to certain restrictions by the
+// United States and other governments.
+// It is your responsibility to ensure that your use and/or transfer does not
+// violate applicable laws.
+// For more information, please see https://www.bis.doc.gov
+// See also https://www.apache.org/dev/crypto.html and/or seek legal counsel.
+func calculateCPUPercent(previous *dtypes.CPUStats, v *dtypes.CPUStats) float64 {
+	var (
+		cpuPercent = 0.0
+		// calculate the change for the cpu usage of the container in between readings
+		cpuDelta = float64(v.CPUUsage.TotalUsage) - float64(previous.CPUUsage.TotalUsage)
+		// calculate the change for the entire system between readings
+		systemDelta = float64(v.SystemUsage) - float64(previous.SystemUsage)
+		onlineCPUs  = float64(v.OnlineCPUs)
+	)
+
+	if onlineCPUs == 0.0 {
+		onlineCPUs = float64(len(v.CPUUsage.PercpuUsage))
+	}
+	if systemDelta > 0.0 && cpuDelta > 0.0 {
+		cpuPercent = (cpuDelta / systemDelta) * onlineCPUs * 100.0
+	}
+	return cpuPercent
+}
+
+func calculateMemoryPercent(memoryStats *dtypes.MemoryStats) float64 {
+	if float64(memoryStats.Limit) == 0 {
+		return 0
+	}
+	return 100.0 * (float64(memoryStats.Usage) - float64(memoryStats.Stats["cache"])) / float64(memoryStats.Limit)
 }
```

Comment:
Would it be worth moving to its own file named `thirdparty.go` just so it is clear on this is the only code copied and modified?

Reply:
This was just moved from `metrics.go` in this PR. Happy to move it elsewhere if required.